File: Proiectarea Algoritmilor/Lab09/codeBase/C++/src-lab6/lab6p1.cpp
#include <iostream>
#include <fstream>
#include <queue>

#include "Maze.h"

typedef std::pair<int, int> Coord;
const Coord NOWHERE(-1, -1);

const int dirLin[4] = { -1, 0, 1, 0 };
const int dirCol[4] = { 0, 1, 0, -1 };

void find_exit(Maze& maze, Coord source)
{
  /* Pentru a reconstitui drumul, vom folosi o matrice de parinti. */
  Coord parent[maze.get_height()][maze.get_width()];
  for (unsigned int i = 0; i < maze.get_height(); ++i) {
    for (unsigned int j = 0; j < maze.get_width(); ++j) {
      parent[i][j] = NOWHERE;
    }
  }

  /* Vom folosi BFS pentru a determina drumul optim. Cream o coada pe care o
   * intializam cu prima celula. */
  std::queue<Coord> q;
  q.push(source);
  parent[source.first][source.second] = source;

  while (!q.empty()) {
    Coord pos = q.front();
    q.pop();

    for (unsigned int dir = 0; dir < 4; ++dir) {
      Coord newPos(pos.first + dirLin[dir], pos.second + dirCol[dir]);
      /* Verificam ca nou pozitie pe care o incercam este o pozitie valida de pe
       * harta in care se poate muta. */
      if (maze.is_walkable(newPos)) {
        if (maze.is_exit_point(newPos)) {
          parent[newPos.first][newPos.second] = pos;
          /* Tiparim solutia si iesim. */
          do {
            maze.mark_solution_step(newPos);
            newPos = parent[newPos.first][newPos.second];
          } while (newPos != source);
          maze.mark_solution_step(source);
          return;
        } else if (parent[newPos.first][newPos.second] == NOWHERE) {
          /* Nu adaugam noua pozitie in coada decat daca nu se afla deja. */
          parent[newPos.first][newPos.second] = pos;
          q.push(newPos);
        }
      }
    }
  }
}

int main()
{
  /* Citim o harta din fisierul de intrare. */
  std::ifstream in("src-lab6/Labirint.txt");
  Maze maze;
  unsigned int lineTrudy, columnTrudy;
  in >> maze >> lineTrudy >> columnTrudy;

  /* Calculam pe ea drumul din labirint. */
  find_exit(maze, Coord(lineTrudy, columnTrudy));

  /* Si afisam drumul final. */
  std::cout << "Labirintul cu drumul marcat spre iesire este: " << std::endl
      << maze;

  return 0;
}

File: Proiectarea Algoritmilor/Lab07/codeBase/Java/src-lab6/Lab6p1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Lab6p1 {

	static final int dirLin[] = { -1, 0, 1, 0 };
	static final int dirCol[] = { 0, 1, 0, -1 };

	public static void find_exit(Maze maze, Coord source) {
		/* Pentru a reconstitui drumul, vom folosi o matrice de parinti. */
		Coord parent[][] = new Coord[maze.get_height()][maze.get_width()];

		for (int i = 0; i < maze.get_height(); ++i) {
			for (int j = 0; j < maze.get_width(); ++j) {
				parent[i][j] = Coord.NOWHERE;
			}
		}

		/*
		 * Vom folosi BFS pentru a determina drumul optim. Cream o coada pe care
		 * o intializam cu prima celula.
		 */
		Queue<Coord> q = new LinkedList<Coord>();

		q.add(source);
		parent[source.lin][source.col] = source;

		while (q.size() > 0) {
			Coord pos = q.remove();

			for (int dir = 0; dir < 4; ++dir) {
				Coord newPos = new Coord(pos.lin + dirLin[dir], pos.col
						+ dirCol[dir]);
				/*
				 * Verificam ca nou pozitie pe care o incercam este o pozitie
				 * valida de pe harta in care se poate muta.
				 */
				if (maze.is_walkable(newPos)) {
					if (maze.is_exit_point(newPos)) {
						parent[newPos.lin][newPos.col] = pos;
						/* Tiparim solutia si iesim. */
						do {
							maze.mark_solution_step(newPos);
							newPos = parent[newPos.lin][newPos.col];
						} while (newPos != source);
						maze.mark_solution_step(source);
						return;
					} else if (parent[newPos.lin][newPos.col] == Coord.NOWHERE) {
						/*
						 * Nu adaugam noua pozitie in coada decat daca nu se
						 * afla deja.
						 */
						parent[newPos.lin][newPos.col] = pos;
						q.add(newPos);
					}
				}
			}
		}
	}

	public static void main(String[] args) throws FileNotFoundException {

		Scanner scanner = new Scanner(new File("Labirint.txt"));

		/* read height, width */
		int height = scanner.nextInt();
		int width = scanner.nextInt();
		scanner.nextLine();

		/* allocate & read maze */
		Maze maze = new Maze(height, width);
		maze.read(scanner);

		/* read robot location */
		int lineTrudy = scanner.nextInt();
		int columnTrudy = scanner.nextInt();
		System.out.println("trudy: " + lineTrudy + " " + columnTrudy);

		/* solve */
		find_exit(maze, new Coord(lineTrudy, columnTrudy));
		
		/* print path */
		maze.print();
	}
}
